redis/src/cluster_slotmap.rs

```diff
@@ -33,41 +29,83 @@ pub(crate) enum ReadFromReplicaStrategy {
 #[derive(Debug, Default)]
 pub(crate) struct SlotMap {
     pub(crate) slots: BTreeMap<u16, SlotMapValue>,
+    pub(crate) nodes_map: NodesMap,
     read_from_replica: ReadFromReplicaStrategy,
 }
 
 fn get_address_from_slot(
     slot: &SlotMapValue,
     read_from_replica: ReadFromReplicaStrategy,
     slot_addr: SlotAddr,
-) -> &str {
-    if slot_addr == SlotAddr::Master || slot.addrs.replicas.is_empty() {
-        return slot.addrs.primary.as_str();
+) -> Arc<String> {
+    let addrs = slot.addrs.read().unwrap();
+    if slot_addr == SlotAddr::Master || addrs.replicas.is_empty() {
+        return addrs.primary.clone();
     }
     match read_from_replica {
-        ReadFromReplicaStrategy::AlwaysFromPrimary => slot.addrs.primary.as_str(),
+        ReadFromReplicaStrategy::AlwaysFromPrimary => addrs.primary.clone(),
         ReadFromReplicaStrategy::RoundRobin => {
             let index = slot
                 .latest_used_replica
                 .fetch_add(1, std::sync::atomic::Ordering::Relaxed)
-                % slot.addrs.replicas.len();
-            slot.addrs.replicas[index].as_str()
+                % addrs.replicas.len();
+            addrs.replicas[index].clone()
         }
     }
 }
 
 impl SlotMap {
     pub(crate) fn new(slots: Vec<Slot>, read_from_replica: ReadFromReplicaStrategy) -> Self {
-        let mut this = Self {
+        let mut slot_map = SlotMap {
             slots: BTreeMap::new(),
+            nodes_map: DashMap::new(),
             read_from_replica,
         };
-        this.slots.extend(
-            slots
-                .into_iter()
-                .map(|slot| (slot.end(), SlotMapValue::from_slot(slot))),
-        );
-        this
+        let mut shard_id = 0;
+        for slot in slots {
+            let primary = Arc::new(slot.master);
+            let replicas: Vec<Arc<String>> = slot.replicas.into_iter().map(Arc::new).collect();
```

Comment:
nit: I would make here `slot.replicas_vec()` method (we might be needing this one-liner elsewhere, so it makes sense to make it a function)

Reply:
we don't use it elsewhere, and using this function partially moves `slot` so we can't consume other properties of this struct (e.g. master)